src/network/https.js

```diff
@@ -139,21 +159,76 @@ class HttpsWorker {
     return Promise.reject(endpoint.onFailure(response));
   }
 
+  /**
+   * Attempts to establish a session for the provided email and password.
+   *
+   * @param {*} email emailed registered by the Wealthsimple Trade account
+   * @param {*} password The password of the account
+   */
+  async login(email, password) {
+    let response = null;
+
+    /*
+     * If we are given a function for otp, then we must fail a log in to
+     * trigger an OTP event with Wealthsimple Trade. This will allow the user
+     * otp thunk to retrieve the code.
+     *
+     * If a literal value is provided for otp, it means the user has manually
+     * provided us with the otp code. We can skip this login attempt.
+     */
+    if (typeof (this.events.otp) === 'function') {
+      await this.handleRequest(endpoints.LOGIN, {
+        email,
+        password,
+      }).catch(() => { });
+    }
+
+    // Try to log in for real this time.
+    try {
+      response = await this.handleRequest(endpoints.LOGIN, {
+        email,
+        password,
+        otp: typeof (this.events.otp) === 'function' ? await this.events.otp() : this.events.otp,
+      });
+    } catch (error) {
+      // we might have failed because OTP was not provided
+      if (!this.events.otp) {
+        throw new Error('OTP not provided!');
+      }
+
+      // Seems to be incorrect credentials or OTP.
+      throw error;
+    }
+
+    // Capture the tokens for later usage.
+    this.tokens.store(response.tokens);
+  }
+
   /**
    * Obtain a new set of authentication tokens using the
    * existing refresh token.
+   *
+   * If we are given a function for refresh, call that function
+   * with the refreshed tokens.
    */
   async refreshAuthentication() {
     const response = await this.handleRequest(endpoints.REFRESH, {
       refresh_token: this.tokens.refresh,
     });
 
     this.tokens.store(response.tokens);
+    // call the handler for this event if we have one
+    if (typeof (this.events.refresh) === 'function') {
+      await this.events.refresh(this.tokens);
+    }
   }
 
   /**
    * Check if the existing set of tokens have expired, automatically
    * triggering a refresh if they have expired.
+   *
```

Comment:
this comment addition can be removed since the `refresh` handler call is no longer happening here too.